// @metorial/anthropic/dist/index.cjs
'use strict';

var mcpSdkUtils = require('@metorial/mcp-sdk-utils');

// src/metorialAnthropic.ts
var metorialAnthropic = mcpSdkUtils.createMcpSdk()(async ({ tools }) => ({
  tools: tools.getTools().map(
    (t) => {
      var _a;
      return {
        name: t.id,
        description: (_a = t.description) != null ? _a : void 0,
        input_schema: t.getParametersAs("json-schema")
      };
    }
  ),
  callTools: async (calls) => ({
    role: "user",
    content: await Promise.all(
      calls.map(async (call) => {
        let tool = tools.getTool(call.name);
        if (!tool) {
          return {
            type: "tool_result",
            tool_use_id: call.id,
            content: `[ERROR] Tool with name "${call.name}" not found.`
          };
        }
        let data = {};
        try {
          if (typeof call.input == "string") {
            data = JSON.parse(call.input);
          } else {
            data = call.input;
          }
        } catch (e) {
          return {
            type: "tool_result",
            tool_use_id: call.id,
            content: `[ERROR] Invalid JSON in tool call arguments: ${e.message}`
          };
        }
        try {
          let result = await tool.call(data);
          return {
            type: "tool_result",
            tool_use_id: call.id,
            content: JSON.stringify(result)
          };
        } catch (e) {
          return {
            type: "tool_result",
            tool_use_id: call.id,
            content: `[ERROR] Tool call failed: ${e.message}`
          };
        }
      })
    )
  })
}));

exports.metorialAnthropic = metorialAnthropic;
