// @metorial/openai-compatible/dist/index.cjs
'use strict';

var mcpSdkUtils = require('@metorial/mcp-sdk-utils');

// src/metorialOpenAICompatible.ts
var createOpenAICompatibleMcpSdk = (opts) => mcpSdkUtils.createMcpSdk()(async ({ tools }) => ({
  tools: tools.getTools().map(
    (t) => {
      var _a;
      return {
        type: "function",
        function: {
          name: t.id,
          description: (_a = t.description) != null ? _a : void 0,
          parameters: t.getParametersAs("json-schema")
        }
      };
    }
  ),
  callTools: async (calls) => Promise.all(
    calls.map(async (call) => {
      let tool = tools.getTool(call.function.name);
      if (!tool) {
        return {
          tool_call_id: call.id,
          role: "tool",
          content: `[ERROR] Tool with name "${call.function.name}" not found.`
        };
      }
      let data = {};
      try {
        if (typeof call.function.arguments == "string") {
          data = JSON.parse(call.function.arguments);
        } else {
          data = call.function.arguments;
        }
      } catch (e) {
        return {
          tool_call_id: call.id,
          role: "tool",
          content: `[ERROR] Invalid JSON in tool call arguments: ${e.message}`
        };
      }
      try {
        let result = await tool.call(data);
        return {
          tool_call_id: call.id,
          role: "tool",
          content: JSON.stringify(result)
        };
      } catch (e) {
        return {
          tool_call_id: call.id,
          role: "tool",
          content: `[ERROR] Tool call failed: ${e.message}`
        };
      }
    })
  )
}));

exports.createOpenAICompatibleMcpSdk = createOpenAICompatibleMcpSdk;
