// @metorial/openai/dist/index.cjs
'use strict';

var mcpSdkUtils = require('@metorial/mcp-sdk-utils');
var openaiCompatible = require('@metorial/openai-compatible');

// src/metorialOpenAI.ts
var metorialOpenAI = {
  chatCompletions: openaiCompatible.createOpenAICompatibleMcpSdk({ withStrict: true }),
  responses: mcpSdkUtils.createMcpSdk()(async ({ session, tools }) => ({
    tools: tools.getTools().map((t) => {
      var _a;
      return {
        type: "function",
        name: t.id,
        description: (_a = t.description) != null ? _a : void 0,
        parameters: t.getParametersAs("json-schema"),
        strict: true
      };
    }),
    callTools: async (calls) => Promise.all(
      calls.map(async (call) => {
        let tool = tools.getTool(call.name);
        if (!tool) {
          return {
            call_id: call.call_id,
            type: "function_call_output",
            output: `[ERROR] Tool with name "${call.name}" not found.`
          };
        }
        let data = {};
        try {
          data = JSON.parse(call.arguments);
        } catch (e) {
          return {
            call_id: call.call_id,
            type: "function_call_output",
            output: `[ERROR] Invalid JSON in tool call arguments: ${e.message}`
          };
        }
        try {
          let result = await tool.call(data);
          return {
            call_id: call.call_id,
            type: "function_call_output",
            output: JSON.stringify(result)
          };
        } catch (e) {
          return {
            call_id: call.call_id,
            type: "function_call_output",
            output: `[ERROR] Tool call failed: ${e.message}`
          };
        }
      })
    )
  }))
};

exports.metorialOpenAI = metorialOpenAI;
