// @metorial/google/dist/index.cjs
'use strict';

var mcpSdkUtils = require('@metorial/mcp-sdk-utils');

// src/metorialGoogle.ts
var metorialGoogle = mcpSdkUtils.createMcpSdk()(async ({ tools }) => ({
  tools: [
    {
      functionDeclarations: tools.getTools().map(
        (t) => {
          var _a;
          return {
            name: t.id,
            description: (_a = t.description) != null ? _a : void 0,
            parameters: t.getParametersAs("openapi-3.0.0")
          };
        }
      )
    }
  ],
  callTools: async (calls) => ({
    role: "user",
    parts: await Promise.all(
      calls.map(async (call) => {
        let tool = tools.getTool(call.name);
        if (!tool) {
          return {
            functionResponse: {
              id: call.id,
              name: call.name,
              response: {
                error: `[ERROR] Tool with name "${call.name}" not found.`
              }
            }
          };
        }
        let data = {};
        try {
          if (typeof call.args == "string") {
            data = JSON.parse(call.args);
          } else {
            data = call.args;
          }
        } catch (e) {
          return {
            functionResponse: {
              id: call.id,
              name: call.name,
              response: {
                error: `[ERROR] Invalid JSON in tool call arguments: ${e.message}`
              }
            }
          };
        }
        try {
          let result = await tool.call(data);
          return {
            functionResponse: {
              id: call.id,
              name: call.name,
              response: result
            }
          };
        } catch (e) {
          return {
            functionResponse: {
              id: call.id,
              name: call.name,
              response: {
                error: `[ERROR] Tool call failed: ${e.message}`
              }
            }
          };
        }
      })
    )
  })
}));

exports.metorialGoogle = metorialGoogle;
