// @metorial/mistral/dist/index.js
import { createMcpSdk } from '@metorial/mcp-sdk-utils';

// src/metorialMistral.ts
var metorialMistral = createMcpSdk()(async ({ tools }) => {
  let toolList = tools.getTools();
  return {
    tools: toolList.map((t) => {
      var _a;
      let parameters = t.getParametersAs("json-schema");
      if (parameters && typeof parameters === "object") {
        parameters.additionalProperties = false;
        if (parameters.properties) {
          Object.values(parameters.properties).forEach((prop) => {
            if (prop && typeof prop === "object" && prop.type === "object") {
              prop.additionalProperties = false;
            }
          });
        }
      }
      return {
        type: "function",
        function: {
          name: t.id,
          description: (_a = t.description) != null ? _a : void 0,
          parameters,
          strict: true
        }
      };
    }),
    callTools: async (calls) => Promise.all(
      calls.map(async (call) => {
        let tool = tools.getTool(call.function.name);
        if (!tool) {
          return {
            toolCallId: call.id,
            role: "tool",
            content: `[ERROR] Tool with name "${call.function.name}" not found.`
          };
        }
        let data = {};
        try {
          if (typeof call.function.arguments == "string") {
            data = JSON.parse(call.function.arguments);
          } else {
            data = call.function.arguments;
          }
        } catch (e) {
          return {
            toolCallId: call.id,
            role: "tool",
            content: `[ERROR] Invalid JSON in tool call arguments: ${e.message}`
          };
        }
        try {
          let result = await tool.call(data);
          return {
            toolCallId: call.id,
            role: "tool",
            content: JSON.stringify(result)
          };
        } catch (e) {
          return {
            toolCallId: call.id,
            role: "tool",
            content: `[ERROR] Tool call failed: ${e.message}`
          };
        }
      })
    )
  };
});

export { metorialMistral };
